src/builtins/read.cpp

```diff
@@ -494,17 +494,22 @@ maybe_t<int> builtin_read(parser_t &parser, io_streams_t &streams, const wchar_t
     do {
         buff.clear();
 
-        // TODO: Determine if the original set of conditions for interactive reads should be
-        // reinstated: if (isatty(0) && streams.stdin_fd == STDIN_FILENO && !split_null) {
         int stream_stdin_is_a_tty = isatty(streams.stdin_fd);
         if (stream_stdin_is_a_tty && !opts.split_null) {
             // Read interactively using reader_readline(). This does not support splitting on null.
             exit_res =
                 read_interactive(parser, buff, opts.nchars, opts.shell, opts.silent, opts.prompt,
                                  opts.right_prompt, opts.commandline, streams.stdin_fd);
         } else if (!opts.nchars && !stream_stdin_is_a_tty &&
-                   lseek(streams.stdin_fd, 0, SEEK_CUR) != -1) {
-            exit_res = read_in_chunks(streams.stdin_fd, buff, opts.split_null);
+                   (streams.stdin_is_directly_redirected || lseek(streams.stdin_fd, 0, SEEK_CUR) != -1)) {
+            // We read in chunks when we either can seek (so we put the bytes back),
+            // or we have the bytes to ourselves (because it's directly redirected).
+            //
+            // Note we skip seeking back even if we're directly redirected to a seekable stream,
+            // under the assumption that the stream will be closed soon anyway.
+            // You don't rewind VHS tapes before throwing them in the trash.
+            // TODO: Do this when nchars is set by seeking back.
+            exit_res = read_in_chunks(streams.stdin_fd, buff, opts.split_null, !streams.stdin_is_directly_redirected);
```

Comment:
Looks like a reasonable change.

> // TODO: Do this when nchars is set by seeking back.

I'm not sure how that will look like. The stream may not be seekable, so we can't actually lseek() back.
But the same assumption holds: we're draining the stream, so it's okay to spill.
Passing nchars to read_in_chunks seems ugly, so  I'd probably just disable this optimization for `read --nchars`.

> Is there a case where we are directly redirected but the fd is used?

Probably not today but in bash we can do something like

	exec 3< README.rst
	read line1 <&3
	read line2 <&3

which would break if read too much.
That's a pretty low level API. In this case it's more elegant than a loop but it's also obscure and easy to misuse.
I don't expect that fish wants to expose fds directly. Would be interesting to know where this feature is used, but I wouldn't care about that now.

Reply:
Okay, I believe this could be triggered via `read <&0` already. I'll have to check later.

Reply:
#5714. Squint and it looks like a feature.

Reply:
What would be a better way to do this without "abusing" exec? I would like to see the feature land here someday but agree maybe something more intuitive could be found.

Reply:
exec --man in ksh:

``` 
  If command is specified, then the current shell process will be replaced by
  command rather than running command and waiting for it to complete. Note that
  there is no need to use exec to enhance performance since the shell
  implicitly uses the exec mechanism internally whenever possible.

If no operands are specified, exec can be used to open or close files, or to
  manipulate file descriptors from 0 to 9 in the current shell environment
  using the standard redirection mechanism available with all commands. The
  close-on-exec flags will be set on file descriptor numbers greater than 2
  that are opened this way so that they will be closed when another program is
  invoked.
```

I think their documentation can be instructive, but my take away is that this other modality under which exec is used, is basically for manipulating fds. I would suggest if we don't ever allow for the second form, we add a builtin `fd` or something.

Reply:
>exec --man in ksh93:

It's *documented*, sure. But how does redirecting arbitrary file descriptors belong to the "make this other command replace me" builtin?

It's two entirely disparate things to do! That's why I call it "abuse". 

But anyway: This has very little to do with `read`, we have other issues open about it.

Reply:
bash's `exec n< foo` can be subbed with `mkfifo` (and I actually have fish scripts that create, pipe to, and read from FIFOs for advanced parallel processing). Would that be an issue here?